ff/src/fields/cyclotomic.rs

```diff
@@ -0,0 +1,123 @@
+/// If a field has multiplicative order p^d - 1, the cyclotomic subgroups refer to subgroups of order φ_n(p),
+/// for any n < d, where φ_n is the [n-th cyclotomic polynomial](https://en.wikipedia.org/wiki/Cyclotomic_polynomial).
+///
+/// ## Note
+///
+/// Note that this trait is unrelated to the `Group` trait from the `ark_ec` crate. That trait
+/// denotes an *additive* group, while this trait denotes a *multiplicative* group.
+pub trait CyclotomicMultSubgroup: crate::Field {
+    /// Is the inverse fast to compute? For example, in quadratic extensions, the inverse
+    /// can be computed at the cost of negating one coordinate, which is much faster than
+    /// standard inversion.
+    /// By default this is `false`, but should be set to `true` for quadratic extensions.
+    const INVERSE_IS_FAST: bool = false;
+
+    /// Compute a square in the cyclotomic subgroup. By default this is computed using [`Field::square`], but for
+    /// degree 12 extensions, this can be computed faster than normal squaring.
+    ///
+    /// # Warning
+    ///
+    /// This method should be invoked only when `self` is in the cyclotomic subgroup.
+    fn cyclotomic_square(&self) -> Self {
+        let mut result = *self;
+        *result.cyclotomic_square_in_place()
+    }
+
+    /// Square `self` in place. By default this is computed using
+    /// [`Field::square_in_place`], but for degree 12 extensions,
+    /// this can be computed faster than normal squaring.
+    ///
+    /// # Warning
+    ///
+    /// This method should be invoked only when `self` is in the cyclotomic subgroup.
+    fn cyclotomic_square_in_place(&mut self) -> &mut Self {
+        self.square_in_place()
+    }
+
+    /// Compute the inverse of `self`. See [`Self::INVERSE_IS_FAST`] for details.
+    /// Returns [`None`] if `self.is_zero()`, and [`Some`] otherwise.
+    ///
+    /// # Warning
+    ///
+    /// This method should be invoked only when `self` is in the cyclotomic subgroup.
+    fn cyclotomic_inverse(&self) -> Option<Self> {
+        let mut result = *self;
+        result.cyclotomic_inverse_in_place().copied()
+    }
+
+    /// Compute the inverse of `self`. See [`Self::INVERSE_IS_FAST`] for details.
+    /// Returns [`None`] if `self.is_zero()`, and [`Some`] otherwise.
+    ///
+    /// # Warning
+    ///
+    /// This method should be invoked only when `self` is in the cyclotomic subgroup.
+    fn cyclotomic_inverse_in_place(&mut self) -> Option<&mut Self> {
+        self.inverse_in_place()
+    }
+
+    /// Compute a cyclotomic exponentiation of `self` with respect to `e`.
+    ///
+    /// # Warning
+    ///
+    /// This method should be invoked only when `self` is in the cyclotomic subgroup.
+    fn cyclotomic_exp(&self, e: impl AsRef<[u64]>) -> Self {
+        let mut result = *self;
+        result.cyclotomic_exp_in_place(e);
+        result
+    }
+
+    /// Set `self` to be the result of exponentiating [`self`] by `e`,
```

Comment:
GitHub said "public documentation for `cyclotomic_exp_in_place` links to private item `self`"